python/lsst/ts/mtdome/mtdome_csc.py

```diff
@@ -321,13 +324,21 @@ async def write_then_read_reply(
         st = encoding_tools.encode(**command_dict)
         async with self.communication_lock:
             self.log.debug(f"Sending command {st}")
-            assert self.writer is not None
-            self.writer.write(st.encode() + b"\r\n")
-            await self.writer.drain()
-            assert self.reader is not None
-            read_bytes = await asyncio.wait_for(
-                self.reader.readuntil(b"\r\n"), timeout=_TIMEOUT
-            )
+            try:
+                assert self.writer is not None
```

Comment:
Please consider checking that you are connected, both here and at line 335 (instead of `assert self.reader is not None`). It is a more useful test: there is no point reading or writing if the connection is gone.

Reply:
In a similar vein, I see that your `connected` property only checks if the reader and writer are None. Please consider something a bit more detailed, such as (from ts_tcpip's OneClientServer):
```
    @property
    def connected(self) -> bool:
        """Return True if a client is connected to this server."""
        return not (
            self.reader is None
            or self.writer is None
            or self.reader.at_eof()
            or self.writer.is_closing()
        )
```

Reply:
I tried that a long time ago but then MyPy isn't happy:

`
329: error: Item "None" of "Optional[StreamWriter]" has no attribute "write"
330: error: Item "None" of "Optional[StreamWriter]" has no attribute "drain"
337: error: Item "None" of "Optional[StreamReader]" has no attribute "readuntil"
`

Reply:
Yes. I know. I use `# type: ignore` to get around that. It's worth it to me to a more complete "is connected" test. But if you really hate `# type: ignore` then you can also assert that the thing is not None. It's just that the full `connected` test is more useful in reality. The other just is a sop to mypy.